<!DOCTYPE html>
<html lang="zh">

<head>
    <meta charset="UTF-8">

    <title>Document</title>
</head>

<body>
    <script>
        /*
        promise对象特点
            1.对象的状态不受外界影响
                pending(进行中),fulfilled(完成),rejected(已失败)
                只有异步操作的结果才能决定当前是哪一种状态,其他任何操作无法改变
            2.状态一旦改变,就不在变.任何时候都会得到这个结果
                从pending变为fulfilled和从pending变为rejected
                只要这两种情况发生，状态就凝固了，不会再变了，会一直保持这个结果，这时就称为 resolved（已定型）
    */

        const promise1 = new Promise((resolve, rejected) => {
            //;;;
            if (1) {//如果异步操作成功
                let value = 'success'
                resolve(value);//返回数据
            } else {

                rejected('error');
            }
        })

        /*
        resolve函数的作用是，将Promise对象的状态从“未完成”变为“成功”（即从 pending 变为 resolved），
        在异步操作成功时调用，并将异步操作的结果，作为参数传递出去；

        reject函数的作用是，将Promise对象的状态从“未完成”变为“失败”（即从 pending 变为 rejected），
        在异步操作失败时调用，并将异步操作报出的错误，作为参数传递出去。
        */


        //promise实例生成后,可以用then指定resolved状态和rejected状态的回调函数。
        promise1.then(value => console.log(value), error => console.log(error))

        //e
        function timeout(ms) {
            return new Promise((resolve, rejected) => {
                setTimeout(resolve, ms, 'done');
            });
        }

        timeout(100).then(value => {
            console.log(value);//done
        })



        //promise新建后就会立即执行
        const promise2 = new Promise((resolve, rejected) => {
            console.log('promise2');
            resolve();
        })

        promise2.then(() => {
            console.log('resolved.');
        });
        console.log('hi');


        //*********promise实例,实现Ajax操作
        const getJSON = function (url) {
            const promise = new Promise((resolve, rejected) => {
                const handler = function () {
                    if (this.readState !== 4)
                        return;
                    if (this.status === 200)
                        resolve(this.response);
                    else
                        rejected(new Error(this.statusText));
                }
                const client = new XMLHttpRequest();
                client.open('GET', url);
                client.onreadystatechange = handler;
                client.responseType = 'json';
                client.setRequestHeader('Accept', 'application/json');
                client.send();
            });

            return promise;
        }

        // getJSON('/posts.json').then(function (json) {
        //     console.log('Contents:' + json);
        // }, function (error) {
        //     console.log('出错了', error);
        // })


        const p1 = new Promise(function (resolve, reject) {
            // ...
        });

        const p2 = new Promise(function (resolve, reject) {
            // ...
            resolve(p1);//即一个异步操作的结果是返回另一个异步操作。
        })
        /*
        注意，这时p1的状态就会传递给p2，也就是说，p1的状态决定了p2的状态。如果p1的状态是pending，
        那么p2的回调函数就会等待p1的状态改变；如果p1的状态已经是resolved或者rejected，
        那么p2的回调函数将会立刻执行。
        */
        new Promise((resolve, reject) => {
            return resolve(1);
            console.log(2);
        }).then(r => {
            console.log(r);
        });
        // 2
        // 1
        //调用resolve(1)以后，后面的console.log(2)还是会执行，并且会首先打印出来。这是因为立即 resolved 的 Promise 
        //是在本轮事件循环的末尾执行，总是晚于本轮循环的同步任务。
        /*
        一般来说，调用resolve或reject以后，Promise 的使命就完成了，后继操作应该放到then方法里面，而不应该直接写在resolve或reject的后面。
        所以，最好在它们前面加上return语句，这样就不会有意外。
        */


        //1.Promise.prototype.then() 
        /*
        then方法第一个参数是resolved状态的回调函数,第二个参数时rejected状态的参数
        then方法返回一个新的Promise对象(不是原来的promise)
        因此可以采用链式写法,即then后面还可以写then
        getJSON("/posts.json").then(function(json) {
              return json.post;
        }).then(function(post) {
                  // ...
        });
        后一个回调函数依据前一个回调的状态来决定是否执行
        */


        //2.Promise.prototype.catch() 
        //Promise.prototype.catch()方法是.then(null, rejection)或.then(undefined, rejection)的别名，用于指定发生错误时的回调函数。

        p.then((val) => console.log('fulfilled:', val))
            .catch((err) => console.log('rejected', err));

        // 等同于
        p.then((val) => console.log('fulfilled:', val))
            .then(null, (err) => console.log("rejected:", err));


        //一般来说，不要在then()方法里面定义 Reject 状态的回调函数（即then的第二个参数），总是使用catch方法。

        //******** bad
        promise
            .then(function (data) {
                // success
            }, function (err) {
                // error
            });

        //********** good
        promise
            .then(function (data) { //cb
                // success
            })
            .catch(function (err) {
                // error
            });


        //3.Promise.prototype.finally() finally()方法用于指定不管 Promise 对象最后状态如何，都会执行的操作
        promise
            .then(result => { ; })
            .catch(error => { ; })
            .finally(() => { ; });
        //finally方法里面的操作，应该是与状态无关的，不依赖于 Promise 的执行结果
        //finally的实现
        Promise.prototype.finally = function (callback) {
            let p = this.constructor;
            return this.then(
                value => p.resolve(callback()).then(() => value),
                reason => p.resolve(callback()).then(() => { throw reason })
            )
        }







        //6.Promise.all()用于将多个 Promise 实例，包装成一个新的 Promise 实例。\
        //const p = Promise.all([p1, p2, p3]);
        //p1,p2,p3全为fulfilled,p才为fulfilled,有一个为rejected,p为rejected
        //这个特性在,数据库中查找多个表时尤其有用(只有当所有的查询都完成,这个总操作才算完成)

        //注意，如果作为参数的 Promise 实例，自己定义了catch方法，那么它一旦被rejected，并不会触发Promise.all()的catch方法。
        const p1 = new Promise((resolve, reject) => {
            resolve('hello');
        })
            .then(result => result)
            .catch(e => e);

        const p2 = new Promise((resolve, reject) => {
            throw new Error('报错了');
        })
            .then(result => result)
            .catch(e => e);

        Promise.all([p1, p2])
            .then(result => console.log(result))
            .catch(e => console.log(e));
        // ["hello", Error: 报错了]

        //如果p2没有自己的catch方法，就会调用Promise.all()的catch方法。




        //*****7.Promise.race()同样是将多个 Promise 实例，包装成一个新的 Promise 实例。
        /*
        const p = Promise.race([p1, p2, p3]);
        只要p1、p2、p3之中有一个实例率先改变状态，p的状态就跟着改变。那个率先改变的 Promise 实例的返回值，就传递给p的回调函数。
        */


        //8.Promise.allSettled()法接受一组 Promise 实例作为参数，包装成一个新的 Promise 实例
        /*
        只有等到所有这些参数实例都返回结果，不管是fulfilled还是rejected，包装实例才会结束。该方法由 ES2020 引入。

        下面代码对服务器发出三个请求，等到三个请求都结束，不管请求成功还是失败，加载的滚动图标就会消失。
        */
        const promises = [
            fetch('/api-1'),
            fetch('/api-2'),
            fetch('/api-3'),
        ];
        //该方法返回的新的 Promise 实例，一旦结束，状态总是fulfilled，不会变成rejected。
        await Promise.allSettled(promises);
        removeLoadingIndicator();



        //9.Promise.any()该方法接受一组 Promise 实例作为参数，包装成一个新的 Promise 实例返回。
        //只要参数实例有一个变成fulfilled状态，包装实例就会变成fulfilled状态；如果所有参数实例都变成rejected状态，包装实例就会变成rejected状态。\





        //10.Promise.resolve() 将现有对象转为 Promise 对象,状态为resolved。
        //  const jsPromise = Promise.resolve($.ajax('/whatever.json'));
        //上面代码将 jQuery 生成的deferred对象，转为一个新的 Promise 对象。
        Promise.resolve('foo')
        // 等价于
        new Promise(resolve => resolve('foo'))



        //11.Promise.reject()会返回一个新的 Promise 实例实例的状态为rejected。




        //12.应用
        //图片加载
        const preloadImage = function (path) {
            return new Promise((resolve, reject) => {
                const imag = new Image();
                image.onload = resolve;
                image.onerror = reject;
                image.src = path;
            });
        };


        //13.Promise.try()
        const f = () => console.log('now');
        Promise.resolve().then(f);
        console.log('next');
        // next
        // now
        //上面讲同步函数f包装成异步函数,所以执行顺序在后面(本轮事件循环的末尾执行。)


        //同步函数同步执行,异步函数异步执行的方法
        //1.async函数
        const f = () => console.log('now');
        (async () => f())()
            .then(() => { console.log('异步') })
        console.log('next');
        // 第二行是一个立即执行的匿名函数，会立即执行里面的async函数，因此如果f是同步的，
        //就会得到同步的结果；如果f是异步的，就可以用then指定下一步，就像上面的写法。
        // now
        // next


        //2.new Promise()写法
        const f = () => console.log('now');
        (
            () => new Promise(
                resolve => resolve(f())
            )
        )();
        console.log('next');
        // now
        // next

        //promise.try()
        const f = () => console.log('now');
        Promise.try(f);
        console.log('next');
        // now
        // next

        
    </script>
</body>

</html>